File: myFinance_improved_V2/lib/features/inventory_analysis/discrepancy/data/datasources/discrepancy_datasource.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import '../models/discrepancy_model.dart';

/// Discrepancy Analytics Remote DataSource
///
/// Handles Supabase RPC calls for inventory discrepancy analysis.
class DiscrepancyDatasource {
  final SupabaseClient _client;

  DiscrepancyDatasource(this._client);

  /// Get discrepancy overview data
  /// RPC: get_discrepancy_overview
  Future<DiscrepancyOverviewModel> getDiscrepancyOverview({
    required String companyId,
    String? period,
  }) async {
    final params = {
      'p_company_id': companyId,
      if (period != null) 'p_period': period,
    };

    final response = await _client
        .rpc<Map<String, dynamic>>('get_discrepancy_overview', params: params)
        .single();

    return _handleResponse(
      response,
      (data) => DiscrepancyOverviewModel.fromJson(data),
    );
  }

  /// Response handler helper
  T _handleResponse<T>(
    Map<String, dynamic> response,
    T Function(Map<String, dynamic>) parser,
  ) {
    // success wrapper pattern handling
    if (response.containsKey('success')) {
      if (response['success'] == true) {
        final data = response['data'];
        if (data is Map<String, dynamic>) {
          return parser(data);
        }
        throw Exception('Invalid data format');
      } else {
        throw Exception(response['error']?.toString() ?? 'Unknown error');
      }
    }

    // direct data pattern
    return parser(response);
  }
}

File: website/src/features/inventory_analysis/data/datasources/DiscrepancyDataSource.ts
/**
 * Discrepancy DataSource
 * Handles Supabase RPC calls for discrepancy overview
 */

import { supabaseService } from '@/core/services/supabase.service';
import {
  DiscrepancyOverview,
  mapDiscrepancyOverviewFromRpc,
} from '../../domain/entities/discrepancyOverview';

export interface GetDiscrepancyOverviewParams {
  companyId: string;
  period?: string;
}

export class DiscrepancyDataSource {
  async getDiscrepancyOverview(params: GetDiscrepancyOverviewParams): Promise<DiscrepancyOverview> {
    const rpcParams: Record<string, unknown> = {
      p_company_id: params.companyId,
    };

    if (params.period) {
      rpcParams.p_period = params.period;
    }

    const data = await supabaseService.rpc<Record<string, unknown>>(
      'get_discrepancy_overview',
      rpcParams
    );

    return mapDiscrepancyOverviewFromRpc(data);
  }
}

// Singleton instance
export const discrepancyDataSource = new DiscrepancyDataSource();
